### ISSUE TITLE
[MacOS] Very low performance with version 1.1.121

### ISSUE THREAD

**Issue by QuantumDeveloper** [KhronosGroup/MoltenVK#742]
Hi,

I am writing my own engine on C# using .Net core 3 and I wrapped macOS window to output my triangle demo there and it actually render, BUT....

I receive only ~140 FPS. I investigate a little and found out that calls to 
`vkCmdEndRenderPass(), vkEndCommandBuffer() and vkQueueSubmi﻿t()`
takes toooo long and I have no idea why, but this is not the end.

Most interesting thing start when I swipe to another virtual screen on my mac and right after that FPS becomes ~5000 (which is actually it should be always)

And when I come back to screen where my window renders, FPS again becomes 120-140.
I dig here a little and found issues #486 and #602 
Seems I faced with the same issues and one more thing I can say that I tested with `presentWithCommandBuffer = false` and `synchronousQueueSubmits = false` and in tyhis case there is no performance improvements, but during window resizing my windows gets closed sporadically and in output I see message that application was exit with code 11. I have no idea what it means, but when both values set to `true`, this happens much less often, but happens anyways sooner or later.

Can you please confirm that I am correct in my assumptions?

**Comment by billhollings** [comment]
@QuantumDeveloper 

There's quite a lot of information here for a simple app, and I'd like to try to narrow down your issues.


>calls to `vkCmdEndRenderPass(), vkEndCommandBuffer() and vkQueueSubmi﻿t()` takes toooo long

How did you trace the timing of these calls. Are you using GPU frame capture...or some other mechanism. And what are you comparing it to?

> when I swipe to another virtual screen on my mac and right after that FPS becomes ~5000

Can you provide further info about what is happening here? Do you mean when your app goes into the background...or no longer renders to a desktop window? Does the GPU that is being used change when this occurs? See the `switchSystemGPU` config value for more info on GPU choices.

The issues in  #486 and #602 are related to semaphore synchronization. You don't indicate that to be a problem...since you focused on the above call durations. However...as discussed in those posts...you can try setting the `MVK_ALLOW_METAL_FENCES` to improve GPU-GPU synchronization efficiency.

The `presentWithCommandBuffer` setting is no longer used...so changing it will have no effect. You should leave `synchronousQueueSubmits` set to `true` unless your app is CPU-bound.

**Comment by QuantumDeveloper** [comment]
@billhollings 
`How did you trace the timing of these calls. Are you using GPU frame capture...or some other mechanism. And what are you comparing it to?`
I mean that calls somehow are slow down rendering cycle and I receive only ~100 FPS which is completely low value for simple demo with one triangle. Completely the same code on Windows is giving 2500 FPS on my very slow home PC. So, obviously something wrong with MoltenVK itself.

`
Can you provide further info about what is happening here? Do you mean when your app goes into the background...or no longer renders to a desktop window? Does the GPU that is being used change when this occurs? See the switchSystemGPU config value for more info on GPU choices.
`
There is no switching to the GPU and the second GPU can give me only 1000 FPS, not more. Also I have no code for switching to any other GPU or something like that. I suspect that when window is not on screen anymore, MoltenVK not using some synchronizations or something like that, which let my rendering cycle work on full speed. Its just a guess.

`try setting the MVK_ALLOW_METAL_FENCES to improve GPU-GPU synchronization efficiency.`
How I can do this from .Net Core? Is there some API available that I can make p/invoke and set this value from outside?

**Comment by billhollings** [comment]
>I mean that calls somehow are slow down rendering cycle and I receive only ~100 FPS which is completely low value for simple demo with one triangle. Completely the same code on Windows is giving 2500 FPS on my very slow home PC. So, obviously something wrong with MoltenVK itself.

If you are actually presenting each frame, macOS limits the FPS to synch with the display. If you are not presenting each frame, then your app can run at full rendering speed. How are you measuring FPS? Are you actually measuring call timings...and if so, how? Or are you just inferring that the calls are slow based on a general measure of FPS?

>>try setting the `MVK_ALLOW_METAL_FENCES` to improve GPU-GPU synchronization efficiency.
How I can do this

All of the MoltenVK configuation parameters can be set via runtime environment variables. Have a look at the documentation comments in `vk_mvk_moltenvk.h`. But...as mentioned above...this will not give you higher FPS is you are presenting each frame.

**Comment by QuantumDeveloper** [comment]
@billhollings 
Yes, I am presenting each frame. I measure FPS as usual - just start high performance timer for 1 second and calculate how much frames I present during that second.
If macOS is limit frames to sync with display, then why my FPS so non-stable like 114, 120, 140, 130, 150 etc? Should it be more stable like 60 60 60 or 120 120 120?

And how often I should present on macOS actually?

Also what about exit code 11 for my app during resizing? Can you say something about that? Could issue with command buffers be the root cause here? I thought it was because of my code, but I could find any issue so I am asking you.

`Have a look at the documentation comments in vk_mvk_moltenvk.h`
I will look at documentation and write here later

**Comment by QuantumDeveloper** [comment]
@billhollings 
OK, to clarify my issue and to be sure we are on the same wave:
I know that macOS and actually Windows will not let me present each frame and frame rate will be maximum of 60 on my PC.
But my concern is that on macOS calls to vulkan functions for end drawing are slow down my render cycle (I suspect that they are waiting to some event to finish on CPU) and all that together slow down the whole app.
For example, I measure each frame time when window is visible and it turns out that time is from 0.005 to 0.02 ms. But when window is not visible (moving to the separate virtual desktop on mac), time of each frame becomes from 0 to 0.001 ms.
**So, I suspect that when window is visible, some calls are blocked on CPU, which causing the whole render loop to wait.**
Obviously this is not correct behaviour at it slowing down the whole app (on windows it behaves in other way). 
And my question is: **could you please say what I must do to make those calls unblock CPU and speedup rendering on macOS?**

**Comment by QuantumDeveloper** [comment]
@billhollings 
I receive crash in libMoltenVK.dylib.
Here is short stacktrace (all I can get throw .net

![image](https://user-images.githubusercontent.com/6085009/66149269-6684de80-e61b-11e9-9dbb-eaa0929ad819.png)


Should I create a separate issue for that?

**Comment by billhollings** [comment]
@QuantumDeveloper 

> Should I create a separate issue for that?

No...it's fine here. You already have a number of issues listed above that we can look at. Some of them might be related.

**Comment by QuantumDeveloper** [comment]
@billhollings 
Ok, good.
Could you please answer something regarding my [previous post]?(https://github.com/KhronosGroup/MoltenVK/issues/742#issuecomment-535885225)

**Comment by billhollings** [comment]
> For example, I measure each frame time when window is visible and it turns out that time is from 0.005 to 0.02 ms. But when window is not visible (moving to the separate virtual desktop on mac), time of each frame becomes from 0 to 0.001 ms.

What is different between these two scenarios? What code is being executed when the desktop is visible vs. not visible?

> And my question is: **could you please say what I must do to make those calls unblock CPU and speedup rendering on macOS?**

Since you are presenting each frame, the OS is involved. It might be that the OS is consuming the 5-20us in order to present your rendered triangle. Unfortunately, I am not able to tell where the delay is occurring without further information.

It's not really an accurate measure to extrapolate the drawing overhead of one triangle. A better measure would be to render an entire scene (or at leasts many, many triangles)...to determine if the MoltenVK environment really is slowing you down, relative to what you are expecting...and experiencing on other platforms.

**Comment by QuantumDeveloper** [comment]
> What is different between these two scenarios? What code is being executed when the desktop is visible vs. not visible?

There is no difference and no any conditions. Exactly the same code executes in both cases. 

> Since you are presenting each frame, the OS is involved. It might be that the OS is consuming the 5-20us in order to present your rendered triangle. Unfortunately, I am not able to tell where the delay is occurring without further information.

If it would be helpful, I can post my BeginDraw and EndDraw code here.

> It's not really an accurate measure to extrapolate the drawing overhead of one triangle. A better measure would be to render an entire scene (or at leasts many, many triangles)...to determine if the MoltenVK environment really is slowing you down, relative to what you are expecting...and experiencing on other platforms.
I agree, but its really strange that I am observing such dramatic difference between those 2 cases.
It would be grate it you can point what I am doing wrong here.

**Comment by QuantumDeveloper** [comment]
@billhollings 
Unfortunately, your fixes to swapchain in 126 version didnt help with my issue and now became event  worth.
FPS left on the same point - 115-120, but now I cannot resize my window. When I try to resize, it crashes with the following message:
`libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument`

**Comment by QuantumDeveloper** [comment]
@billhollings 
After detailed analysis I found that crash during resizing was because of my code (at least after fix I cannot repro this), **BUT** FPS is still critically low. **120 FPS** in windowed mode and **10 FPS** in full screen mode. This was tested on MoltenVK 1.1.126.

**Comment by dimi309** [comment]
I too have noticed a very significant drop in framerate when going full screen on catalina on a 2017 macbook pro (13 inch non-retina display), using the .126 version of the vulkan/moltenvk distribution. My programs are using the glfw library. I have tried to unset the MVK_CONFIG_SYNCHRONOUS_QUEUE_SUBMITS environment variable as suggested in one discussion I found online, but it does not seem to have any effect. I am not sure I have done that correctly by the way. I have tried to export it as 0, "false" and "FALSE". I have also been tried to set synchronousQueueSubmits in vk_mvk_moltenvk.h to false but I am not quite sure how to manipulate those settings correctly. Just including vk_mvk_moltenvk.h the file in my program results in unresolved symbols, which is corrected if I link to libMoltenVK.dylib but then I get a segmentation fault, I suppose because the vulkan dylib is also referencing it.

**Comment by billhollings** [comment]
Duplicate of #808?

**Comment by billhollings** [comment]
@QuantumDeveloper @dimi309 

Possible duplicate of #808?

Can you confirm that you were also using 2 swapchain images and that moving to 3 swapchain images fixes this?

**Comment by QuantumDeveloper** [comment]
Duplicate my answer here.

@billhollings yes, I also used 2 images before, but I cannot say that changing it to 3 fixes an issue. It definitely increase from 5-10 to 2500 FPS in full screen mode, BUT at the same time in windowed mode it still stays nearly 250-300 FPS, which is obviously not correct.
FPS count should be the same independently from window mode

**Comment by dimi309** [comment]
Indeed! I was using double buffering and with triple buffering the problem has been resolved. Thanks a lot!
